runtime/service/src/main/java/org/apache/polaris/service/events/listeners/InMemoryBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,195 @@
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(InMemoryBufferPolarisPersistenceEventListener.class);
+  private static final String REQUEST_ID_KEY = "requestId";
+  private final MetaStoreManagerFactory metaStoreManagerFactory;
+
+  private final ConcurrentHashMap<String, ConcurrentLinkedQueueWithApproximateSize<PolarisEvent>>
+      buffer = new ConcurrentHashMap<>();
+  private final ScheduledExecutorService executor;
+  private final Set<Future<?>> futures = ConcurrentHashMap.newKeySet();
+  private final Duration timeToFlush;
+  private final int maxBufferSize;
+
+  @Inject CallContext callContext;
+  @Inject Clock clock;
+  @Context SecurityContext securityContext;
+  @Context ContainerRequestContext containerRequestContext;
+  @Inject PolarisDiagnostics polarisDiagnostics;
+
+  @Inject
+  public InMemoryBufferPolarisPersistenceEventListener(
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      Clock clock,
+      InMemoryBufferEventListenerConfiguration eventListenerConfiguration) {
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.clock = clock;
+    this.timeToFlush = eventListenerConfiguration.bufferTime();
+    this.maxBufferSize = eventListenerConfiguration.maxBufferSize();
+
+    executor = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  @PostConstruct
+  void start() {
+    futures.add(
+        executor.scheduleAtFixedRate(
+            this::runCleanup, 0, timeToFlush.toMillis(), TimeUnit.MILLISECONDS));
+  }
+
+  void runCleanup() {
+    for (String realmId : buffer.keySet()) {
+      try {
+        checkAndFlushBufferIfNecessary(realmId, false);
+      } catch (Exception e) {
+        LOGGER.debug("Buffer checking task failed for realm ({}): {}", realmId, e);
+      }
+    }
+    // Clean up futures
+    try {
+      futures.removeIf(Future::isDone);
+    } catch (Exception e) {
+      LOGGER.debug("Futures reaper task failed.");
+    }
+  }
+
+  @PreDestroy
+  void shutdown() {
+    futures.forEach(future -> future.cancel(false));
+    executor.shutdown();
+
+    try {
+      if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
+        executor.shutdownNow();
+        if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
+          LOGGER.warn("Executor did not shut down cleanly");
+        }
+      }
+    } catch (InterruptedException e) {
+      executor.shutdownNow();
+      Thread.currentThread().interrupt();
+    } finally {
+      for (String realmId : buffer.keySet()) {
+        try {
+          checkAndFlushBufferIfNecessary(realmId, true);
+        } catch (Exception e) {
+          LOGGER.debug("Buffer flushing task failed for realm ({}): ", realmId, e);
+        }
+      }
+    }
+  }
+
+  @Override
+  String getRequestId() {
+    if (containerRequestContext != null && containerRequestContext.hasProperty(REQUEST_ID_KEY)) {
+      return (String) containerRequestContext.getProperty(REQUEST_ID_KEY);
+    }
+    return UUID.randomUUID().toString();
```

Comment:
Minor: having a fake request id may cause issue when users are trying to debug with it, as it doesn't exist anywhere else. I think the containerRequestContext mighty always have one. At least, it's more of the containerRequestContext's responsibility to generate one if it's missing. For event, we could just use what containerRequestContext has.